File: android-reactive-location/src/main/java/pl/charmas/android/reactivelocation2/observables/ObservableFactory.java
package pl.charmas.android.reactivelocation2.observables;


import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.core.ObservableOnSubscribe;
import io.reactivex.rxjava3.core.ObservableSource;
import io.reactivex.rxjava3.core.ObservableTransformer;
import io.reactivex.rxjava3.functions.BiPredicate;

public class ObservableFactory {
    private final ObservableContext context;

    public ObservableFactory(ObservableContext context) {
        this.context = context;
    }

    public <T> Observable<T> createObservable(ObservableOnSubscribe<T> source) {
        return Observable.create(source).compose(new RetryOnConnectionSuspension<T>(context.isRetryOnConnectionSuspended()));
    }

    private static class RetryOnConnectionSuspension<T> implements ObservableTransformer<T, T> {
        private final boolean shouldRetry;

        RetryOnConnectionSuspension(boolean shouldRetry) {
            this.shouldRetry = shouldRetry;
        }

        @Override
        public ObservableSource<T> apply(Observable<T> upstream) {
            if (shouldRetry) {
                return upstream.retry(new IsConnectionSuspendedException());
            }
            return upstream;
        }

        private static class IsConnectionSuspendedException implements BiPredicate<Integer, Throwable> {
            @Override
            public boolean test(Integer integer, Throwable throwable) throws Exception {
                return throwable instanceof GoogleAPIConnectionSuspendedException;
            }
        }
    }
}

File: android-reactive-location/src/main/java/pl/charmas/android/reactivelocation2/observables/ObservableFactory.kt
package pl.charmas.android.reactivelocation2.observables

import io.reactivex.Observable
import io.reactivex.ObservableOnSubscribe
import io.reactivex.ObservableSource
import io.reactivex.ObservableTransformer
import io.reactivex.functions.BiPredicate
import pl.charmas.android.reactivelocation2.observables.exceptions.GoogleAPIConnectionSuspendedException

class ObservableFactory constructor(private val context: ObservableContext) {
    fun <T> createObservable(source: ObservableOnSubscribe<T>): Observable<T> {
        return Observable.create(source).compose(
            RetryOnConnectionSuspension(
                context.isRetryOnConnectionSuspended
            )
        )
    }

    private class RetryOnConnectionSuspension<T> internal constructor(private val shouldRetry: Boolean) :
        ObservableTransformer<T, T> {
        override fun apply(upstream: Observable<T>): ObservableSource<T> {
            return if (shouldRetry) {
                upstream.retry(IsConnectionSuspendedException())
            } else upstream
        }

        private class IsConnectionSuspendedException :
            BiPredicate<Int, Throwable> {
            @Throws(Exception::class)
            override fun test(integer: Int, throwable: Throwable): Boolean {
                return throwable is GoogleAPIConnectionSuspendedException
            }
        }
    }
}